### PR TITLE
perf(ft-asm): express coverage through shortest path on graph

### PR DESCRIPTION
**Pull request by keynslug**
The current algorithm's complexity proved to be unacceptable with sufficiently large inputs (e.g. GiBs-worth files split into 100 KiB sized segments for example). Express and reimplement it as the shortest path problem on graph where nodes are offsets and edges are Erlang nodes' segments.

---

[EMQX-8916](https://emqx.atlassian.net/browse/EMQX-8916)

### CODE REVIEW

File: apps/emqx_ft/src/emqx_ft_assembly.erl [hunk 1]
```diff
@@ -137,42 +137,160 @@ append_filemeta(Asm, Node, Fragment = #{fragment := {filemeta, Meta}}) ->
...
         % be able to disambiguate them though.
-        segs = orddict:store({Offset, locality(Node), -End, Node}, Fragment, Asm#asm.segs)
+        segs = Segs
     }.
 
-coverage([{{Offset, _, _, _}, _Segment} | Rest], Cursor, Sz) when Offset < Cursor ->
-    coverage(Rest, Cursor, Sz);
-coverage([{{Cursor, _Locality, MEnd, Node}, Segment} | Rest], Cursor, Sz) ->
+coverage(Segs, Size) ->
+    find_shortest_path(Segs, 0, Size).
```
Comment (savonarola) [inline apps/emqx_ft/src/emqx_ft_assembly.erl:158]: Could we use https://www.erlang.org/doc/man/digraph.html#get_short_path-3?

File: apps/emqx_ft/src/emqx_ft_assembly.erl [hunk 2]
```diff
@@ -137,42 +137,160 @@ append_filemeta(Asm, Node, Fragment = #{fragment := {filemeta, Meta}}) ->
...
+
+find_shortest_path(G1, From, To) ->
```
Comment (savonarola) [inline apps/emqx_ft/src/emqx_ft_assembly.erl:160]: Graph-related code seems to be quite unrelated to the assembly domain 🤔 
Maybe we move it into a separate module?

Also it seems to be quite important to describe how coverage task domain maps to the graph domain (what are vertices, edges, and how path is related to the coverage)

**Comment by savonarola** [comment]
Great application of the graph algorithm 👍

File: apps/emqx_ft/src/emqx_ft_assembly.erl (hunk 1 above)
Comment (keynslug) [reply apps/emqx_ft/src/emqx_ft_assembly.erl:158]: Tried it initially, was not happy with the performance, not much better than our naive algorithm. Not sure why, haven't looked closely into the implementation.

Comment (keynslug) [reply apps/emqx_ft/src/emqx_ft_assembly.erl:158]: Also it has no concept of cost / weights, though that was easy to overcome with extra nodes.

**Review by savonarola — APPROVED** [review]
